### ISSUE TITLE
minishift fails to find hyper-v virtual switch on windows with polish locale

### ISSUE THREAD

**Issue by kamkie** [minishift/minishift#2096]
### General information

  * Minishift version: 
  * OS: Windows
  * Hypervisor: Hyper-V

### Steps to reproduce

  1. have windows with Polish locale
  2.  set HYPERV_VIRTUAL_SWITCH=Przełącznik domyślny
  3. minishift start
  4. 

### Expected
minishift starts

### Actual
minishift fails to find virtual switch because there is problem with encoding

### Logs
Error starting the VM: Error creating the VM. Error with pre-create check: "vswitch \"Prze\\u0088¥cznik domy\\u0098lny\" not found"

**Comment by kamkie** [comment]
@kev5 your problem looks differently 
`set HYPERV_VIRTUAL_SWITCH="External VM Switch"`
this should fix it

**Comment by kev5** [comment]
Nope, tried it already, doesn't fix it.

**Comment by kamkie** [comment]
@kev5 
set HYPERV_VIRTUAL_SWITCH=External VM Switch
quote is included in cmd
for bash you need them, but in cmd not

**Comment by kamkie** [comment]
using DockerNAT works on that machine but  DockerNAT has its own problems 
tomorrow i will try to test different encodings, maybe setting console to UTF-8

**Comment by gbraad** [comment]
@agajdosi we really have to look into automated tests on localized versions of Windows. This is another good example where stuff goes haywire. What can I do to help with this?

**Comment by kamkie** [comment]
@gbraad enabling nat for DockerNAT is my current workaround https://github.com/kamkie/kamkie-infra/blob/master/minishift/minishift-start-windows-hyperv.ps1

**Comment by kev5** [comment]
Sorry for posting my issue here. Thought it was related. But yeah, I got it working now. Thanks a lot.

**Comment by agajdosi** [comment]
@gbraad True. I have it in my mind, but did not yet progressed due to the unstability of the central-ci slaves. However will check the possibilities on the image I use and tomorrow check with guys from DevStudio QE what are the possibilities.

**Comment by gbraad** [comment]
We need to be able to switch locality between some or before all of the runs. Maybe better to make this part of nightly

**Comment by kamkie** [comment]
@gbraad "Przełącznik domyślny" is "Default Switch" also "Hyper-V Administrators" group is named differently on non English windows  

I tried different encodings and different shells but was unable to find any workarounds 
what works best is set HYPERV_VIRTUAL_SWITCH to DockerNAT  and using managment console switch network adapter to "Default Switch" seconds after vm is created

**Comment by gbraad** [comment]
@kamkie what do you get as result on the command:

```
PS> (Hyper-V\Get-VMSwitch).Name
```

and

```
PS> (Hyper-V\Get-VMSwitch -Name "Przełącznik domyślny").Name
```

**Comment by kamkie** [comment]
@gbraad 
```
PS C:\WINDOWS\system32> (Hyper-V\Get-VMSwitch).Name
Przełącznik domyślny
DockerNAT
PS C:\WINDOWS\system32> (Hyper-V\Get-VMSwitch -Name "Przełącznik domyślny").Name
Przełącznik domyślny
```

**Comment by kamkie** [comment]
encoding problem in go environment variable handling?

**Comment by gbraad** [comment]
Likely, and trying to pinpoint... but golang itself is unicode-aware AFAIR:
https://play.golang.org/p/BI4OAa1hlRQ

**Comment by gbraad** [comment]
Not sure if it is even the environment variable handling. At least on
Linux it seems work: https://play.golang.org/p/ICyRxO9mvfg Wil also
run this from powershell

On Tue, Apr 10, 2018 at 6:38 PM, Gerard Braad <me@gbraad.nl> wrote:
> Likely, and trying to pinpoint... but golang itself is unicode-aware AFAIR:
> https://play.golang.org/p/BI4OAa1hlRQ



-- 

   Gerard Braad | http://gbraad.nl
   [ Doing Open Source Matters ]

**Comment by kamkie** [comment]
I found mentions of that problem in python and ruby  
https://measureofchaos.wordpress.com/2011/03/04/python-on-windows-unicode-environment-variables/comment-page-1/
https://bugs.ruby-lang.org/issues/5570

**Comment by gbraad** [comment]
Interesting... thanks! so it is only windows specific...
will try to reproduce this now... switching systems.

**Comment by gbraad** [comment]
On Windows I get the following

C:\Users\gbraad>env.exe
'Przelacznik domyslny'

PS C:\Users\gbraad> .\env.exe
'Przelacznik domyslny'

So it strips ANY specific character from os.Setenv and os.Getenv

On Tue, Apr 10, 2018 at 7:10 PM, Gerard Braad <me@gbraad.nl> wrote:
> Interesting... thanks! so it is only windows specific...
> will try to reproduce this now... switching systems.



-- 

   Gerard Braad | http://gbraad.nl
   [ Doing Open Source Matters ]

**Comment by gbraad** [comment]
Think we have to resolve this by picking up from this discussion https://github.com/minishift/minishift/issues/1622#issuecomment-361563276 We wanted to create a config option to handle this for all profiles, but due to some time constraints or whatever (unable to remember) we chose not to. the issue got closed with a partial implementation... and seems we forgot to create a new issue. But this could work as a workaround, so will look into this...

**Comment by gbraad** [comment]
Will work on #2258 and see if this might fix the problem as a workaround. But for me, this is also the better approach to handle setting the vswitch.

**Comment by gbraad** [comment]
Might consider to also add a windows specific environment variables handling as part of our utils using `_windows.go`

**Comment by gbraad** [comment]
```
PS C:\work\src\github.com\minishift\minishift> .\out\windows-amd64\minishift.exe config set hyperv-virtual-switch $MV
S
PS C:\work\src\github.com\minishift\minishift> .\out\windows-amd64\minishift.exe config view
- disk-size            : 5GB
- hyperv-virtual-switch: Przełącznik domyślny
- iso-url              : centos
- memory               : 2GB
- vm-driver            : hyperv
PS C:\work\src\github.com\minishift\minishift> .\out\windows-amd64\minishift.exe start --no-provision
-- Starting profile 'minishift'
-- Checking if https://github.com is reachable (using proxy: "No") ... OK
-- Checking if requested OpenShift version 'v3.9.0' is valid ... OK
-- Checking if requested OpenShift version 'v3.9.0' is supported ... OK
-- Checking if requested hypervisor 'hyperv' is supported on this platform ... OK
-- Checking if Hyper-V driver is installed ... OK
-- Checking if Hyper-V driver is configured to use a Virtual Switch ...
   'Przełącznik domyślny' ... OK
-- Checking if user is a member of the Hyper-V Administrators group ... OK
-- Checking the ISO URL ... OK
-- Checking if provided oc flags are supported ... OK
-- Starting local OpenShift cluster using 'hyperv' hypervisor ...
-- Starting Minishift VM .... FAIL E0416 16:22:38.534441   14196 start.go:370] Error starting the VM: Error starting stopped host: exit status 1. Retrying.
```


hahahahaa... it actually works on our end... but fails for Libmachine... You can try. I think for you it will work.

Our check passes the default switch selection, as it seems to check for a localized name, but then it would fail in libmachine on start. However, this is not a big deal... it will actually work on a localized OS.

**Comment by kamkie** [comment]
@gbraad I need to build minishift from master?

**Comment by gbraad** [comment]
@kamkie you can get a binary from:
http://artifacts.ci.centos.org/minishift/minishift/pr/2291/

**Comment by gbraad** [comment]
@kamkie any possibility to have a look?

**Comment by kamkie** [comment]
I currently don't have computers to check that (my work and personal laptop are set to English locale) 
Maybe i will able to find some spare laptop at work tomorrow

**Comment by anjannath** [comment]
This can be closed too as we have #2429 and #2291 is merged now.

**Comment by gbraad** [comment]
https://github.com/docker/machine/pull/4509

**Comment by anjannath** [comment]
Upstream PR https://github.com/machine-drivers/machine/pull/10

**Comment by pablonautic** [comment]
Hi guys,

in case someone follows my path, I will just leave a quick note - I just landed here trying to solve my basic problem with running minishift:

```
C:\Users\pawel\Downloads\minishift-1.28.0-windows-amd64>minishift start --hyperv-virtual-switch=external-vswitch
-- minishift version: v1.28.0+48e89ed
-- Starting profile 'minishift'
-- Check if deprecated options are used ... OK
-- Checking if https://github.com is reachable ... OK
-- Checking if requested OpenShift version 'v3.11.0' is valid ... OK
-- Checking if requested OpenShift version 'v3.11.0' is supported ... OK
-- Checking if requested hypervisor 'hyperv' is supported on this platform ... OK
-- Checking if Hyper-V driver is installed ... OK
-- Checking if Hyper-V driver is configured to use a Virtual Switch ...
   'external-vswitch' ... FAIL
   See the 'Setting Up the Virtualization Environment' topic (https://docs.okd.io/latest/minishift/getting-started/setting-up-virtualization-environment.html) for more information
```

The switch name and everything was fine, so after trying several possible solutions, I run:

`C:\Users\pawel\Downloads\minishift-1.28.0-windows-amd64>minishift config set hyperv-virtual-switch external-vswitch`

With the output:
`[Virtual Switch not found]`

So this error message was enough to hit me - I was not using the elevated (Admin) console! 

Not sure how deep this error message is hidden, but probably it would be worth improving, if possible.

Best regards :)

**Comment by gbraad** [comment]
Can you use PowerShell and confirm the output with:

```
Get-VMSwitch
```

**Comment by dickensas** [comment]
Launching the command prompt in administrator mode solves the problem

```
minishift config set hyperv-virtual-switch "External VM Switch"
```
